### ISSUE TITLE
Gateway cannot be called using localhost when running in non-local silo mode.  Kubernetes pod scenerio

### ISSUE THREAD

**Issue by jms69** [dotnet/orleans#3984]
This is the scenario that breaks:

- Orleans silo running in container listening on port 22222 and gateway 40000 with IP.Any using Azure table membership
- Web app in container listening to port 80 using Orleans client calling Gateway client endpoint 127.0.0.1 port 40000
- Both containers are in side one Kubernetes Statefulset pod with external port 80 exposed.
- Silo to silo communications works well.

It is expected that the web app can communicate with the Orleans silo at localhost:40000 inside the pod.  This is not the case.  It seems that the gateway will always listen on the same address as the silo address.

```
this.gatewayAddressLazy = new Lazy<SiloAddress>(() => endpointOptions.ProxyPort != 0 ? SiloAddress.New(new IPEndPoint(this.SiloAddress.Endpoint.Address, endpointOptions.ProxyPort), 0) : null);
```
The config setting below where IP address is set to ANY is not being honored:
```
config.Defaults.ProxyGatewayEndpoint = new IPEndPoint(IPAddress.Any, 40000);
```

This breaks any sidecar scenarios in containerized applications.

**Comment by jms69** [comment]
Proposed fix: https://github.com/dotnet/orleans/pull/3986

**Comment by jms69** [comment]
Seems like the same issue as https://github.com/dotnet/orleans/issues/3676

I was trying to add an envoy proxy side car to the gateway  port when I discovered this problem.  I was experimenting on possibly securing silo communication with the envoy proxy.

https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/ssl.html?highlight=tls

**Comment by galvesribeiro** [comment]
@jms69 looks like a good idea. Was discussing reverse proxy options last week. Will have a look on Envoy.

**Comment by sergeybykov** [comment]
Resolved via #4005.
